Diary export not working on Android

It seems that exporting diaries is currently broken on Android. After clicking the export button (and having granted file system permissions) a red box is shown at the bottom, saying ```FILE_NOTCREATED```.

Logs:

```
2022-07-20 10:12:30.009 2309-2893/de.ifgi.simport.learning V/Capacitor/Plugin: To native (Capacitor plugin): callbackId: 26969947, pluginId: CapacitorSQLite, methodName: query
2022-07-20 10:12:30.010 2309-2893/de.ifgi.simport.learning V/Capacitor: callback: 26969947, pluginId: CapacitorSQLite, methodName: query, methodData: {"database":"trajectories","statement":"SELECT * FROM diaryEntry","values":[]}
2022-07-20 10:12:30.014 2309-2893/de.ifgi.simport.learning V/Capacitor/Plugin: To native (Capacitor plugin): callbackId: 26969948, pluginId: CapacitorSQLite, methodName: query
2022-07-20 10:12:30.014 2309-2893/de.ifgi.simport.learning V/Capacitor: callback: 26969948, pluginId: CapacitorSQLite, methodName: query, methodData: {"database":"trajectories","statement":"SELECT * FROM logs","values":[]}
2022-07-20 10:12:30.040 2309-2893/de.ifgi.simport.learning V/Capacitor/Plugin: To native (Capacitor plugin): callbackId: 26969949, pluginId: Filesystem, methodName: writeFile
2022-07-20 10:12:30.040 2309-2893/de.ifgi.simport.learning V/Capacitor: callback: 26969949, pluginId: Filesystem, methodName: writeFile, methodData: {"data":"UEsDBAoAAAAAAI9B9FRZ1cVpMAAAADAAAAAdAAAAZGlhcnlfMjAyMi0wNy0yMFQwOC0xMi0zMC5jc3ZkYXRlLGNvbnRlbnQKMjAyMi0wNy0yMFQwNzo1MjoxMS4wMDBaLGdoampoZ2dqaApQSwMECgAAAAAAj0H0VBu6C65EAAAARAAAABsAAABsb2dfMjAyMi0wNy0yMFQwOC0xMi0zMC5jc3Z0eXBlLHNjb3BlLGxldmVsLHRleHQsdGltZXN0YW1wLGxvY2F0aW9uQ291bnQsbGFzdExvY2F0aW9uVGltZXN0YW1wClBLAQIUAAoAAAAAAI9B9FRZ1cVpMAAAADAAAAAdAAAAAAAAAAAAAAAAAAAAAABkaWFyeV8yMDIyLTA3LTIwVDA4LTEyLTMwLmNzdlBLAQIUAAoAAAAAAI9B9FQbuguuRAAAAEQAAAAbAAAAAAAAAAAAAAAAAGsAAABsb2dfMjAyMi0wNy0yMFQwOC0xMi0zMC5jc3ZQSwUGAAAAAAIAAgCUAAAA6AAAAAAA","path":"SIMPORT_export_2022-07-20T08-12-30.zip","directory":"EXTERNAL_STORAGE"}
2022-07-20 10:12:30.041 2309-2874/de.ifgi.simport.learning V/Capacitor/Filesystem: Permission 'android.permission.WRITE_EXTERNAL_STORAGE' is granted
2022-07-20 10:12:30.048 2309-2874/de.ifgi.simport.learning E/Capacitor/Filesystem: Creating binary file '/storage/emulated/0/SIMPORT_export_2022-07-20T08-12-30.zip' failed. Error: /storage/emulated/0/SIMPORT_export_2022-07-20T08-12-30.zip: open failed: EPERM (Operation not permitted)
    java.io.FileNotFoundException: /storage/emulated/0/SIMPORT_export_2022-07-20T08-12-30.zip: open failed: EPERM (Operation not permitted)
        at libcore.io.IoBridge.open(IoBridge.java:492)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:236)
        at com.getcapacitor.plugin.Filesystem.saveFile(Filesystem.java:281)
        at com.getcapacitor.plugin.Filesystem.writeFile(Filesystem.java:225)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:99)
        at com.getcapacitor.Bridge$1.run(Bridge.java:542)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.os.HandlerThread.run(HandlerThread.java:67)
     Caused by: android.system.ErrnoException: open failed: EPERM (Operation not permitted)
        at libcore.io.Linux.open(Native Method)
        at libcore.io.ForwardingOs.open(ForwardingOs.java:166)
        at libcore.io.BlockGuardOs.open(BlockGuardOs.java:254)
        at libcore.io.ForwardingOs.open(ForwardingOs.java:166)
        at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:7542)
        at libcore.io.IoBridge.open(IoBridge.java:478)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:236) 
        at com.getcapacitor.plugin.Filesystem.saveFile(Filesystem.java:281) 
        at com.getcapacitor.plugin.Filesystem.writeFile(Filesystem.java:225) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.getcapacitor.PluginHandle.invoke(PluginHandle.java:99) 
        at com.getcapacitor.Bridge$1.run(Bridge.java:542) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
2022-07-20 10:12:30.049 2309-2874/de.ifgi.simport.learning D/Capacitor: Sending plugin error: {"save":false,"callbackId":"26969949","pluginId":"Filesystem","methodName":"writeFile","success":false,"error":{"message":"FILE_NOTCREATED"}}
```

--- a/android/app/src/main/AndroidManifest.xml
+++ b/android/app/src/main/AndroidManifest.xml
@@ -28,6 +28,7 @@
             android:name="de.ifgi.simport.learning.MainActivity"
             android:label="@string/title_activity_main"
             android:theme="@style/AppTheme.NoActionBarLaunch"
+            android:exported="true"
             android:launchMode="singleTask">
 
             <intent-filter>

--- a/android/app/src/main/java/de/ifgi/simport/learning/MainActivity.java
+++ b/android/app/src/main/java/de/ifgi/simport/learning/MainActivity.java
@@ -6,6 +6,7 @@ import com.bkon.capacitor.fileselector.FileSelector;
 import com.getcapacitor.BridgeActivity;
 import com.getcapacitor.Plugin;
 import com.getcapacitor.community.database.sqlite.CapacitorSQLite;
+import com.getcapacitor.plugin.Filesystem;
 
 import java.util.ArrayList;
 
@@ -20,6 +21,7 @@ public class MainActivity extends BridgeActivity {
       // Ex: add(TotallyAwesomePlugin.class);
       add(CapacitorSQLite.class);
       add(FileSelector.class);
+      add(Filesystem.class);
     }});
   }
 }

--- a/src/app/diary/diary-detail/diary-detail.component.spec.ts
+++ b/src/app/diary/diary-detail/diary-detail.component.spec.ts
@@ -2,6 +2,7 @@ import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing'
 import { APP_TEST_IMPORTS } from 'src/app/app.declarations'
 import { SqliteService } from 'src/app/shared-services/db/sqlite.service'
 import { DiaryService } from 'src/app/shared-services/diary/diary.service'
+import { TrajectoryService } from 'src/app/shared-services/trajectory/trajectory.service'
 
 import { DiaryDetailComponent } from './diary-detail.component'
 
@@ -9,19 +10,17 @@ describe('DiaryDetailComponent', () => {
   let component: DiaryDetailComponent
   let fixture: ComponentFixture<DiaryDetailComponent>
 
-  beforeEach(
-    waitForAsync(() => {
-      TestBed.configureTestingModule({
-        declarations: [DiaryDetailComponent],
-        imports: APP_TEST_IMPORTS,
-        providers: [DiaryService, SqliteService],
-      }).compileComponents()
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      declarations: [DiaryDetailComponent],
+      imports: APP_TEST_IMPORTS,
+      providers: [DiaryService, TrajectoryService, SqliteService],
+    }).compileComponents()
 
-      fixture = TestBed.createComponent(DiaryDetailComponent)
-      component = fixture.componentInstance
-      fixture.detectChanges()
-    })
-  )
+    fixture = TestBed.createComponent(DiaryDetailComponent)
+    component = fixture.componentInstance
+    fixture.detectChanges()
+  }))
 
   it('should create', () => {
     expect(component).toBeTruthy()

--- a/src/app/diary/diary-edit/diary-edit.component.spec.ts
+++ b/src/app/diary/diary-edit/diary-edit.component.spec.ts
@@ -2,6 +2,7 @@ import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing'
 import { APP_TEST_IMPORTS } from 'src/app/app.declarations'
 import { SqliteService } from 'src/app/shared-services/db/sqlite.service'
 import { DiaryService } from 'src/app/shared-services/diary/diary.service'
+import { TrajectoryService } from 'src/app/shared-services/trajectory/trajectory.service'
 
 import { DiaryEditComponent } from './diary-edit.component'
 
@@ -9,19 +10,17 @@ describe('DiaryEditComponent', () => {
   let component: DiaryEditComponent
   let fixture: ComponentFixture<DiaryEditComponent>
 
-  beforeEach(
-    waitForAsync(() => {
-      TestBed.configureTestingModule({
-        declarations: [DiaryEditComponent],
-        providers: [DiaryService, SqliteService],
-        imports: APP_TEST_IMPORTS,
-      }).compileComponents()
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      declarations: [DiaryEditComponent],
+      providers: [DiaryService, TrajectoryService, SqliteService],
+      imports: APP_TEST_IMPORTS,
+    }).compileComponents()
 
-      fixture = TestBed.createComponent(DiaryEditComponent)
-      component = fixture.componentInstance
-      fixture.detectChanges()
-    })
-  )
+    fixture = TestBed.createComponent(DiaryEditComponent)
+    component = fixture.componentInstance
+    fixture.detectChanges()
+  }))
 
   it('should create', () => {
     expect(component).toBeTruthy()

--- a/src/app/diary/diary.page.ts
+++ b/src/app/diary/diary.page.ts
@@ -1,15 +1,8 @@
 import { Component, OnInit } from '@angular/core'
 import { Router } from '@angular/router'
-import { Platform, ToastController } from '@ionic/angular'
-import { TranslateService } from '@ngx-translate/core'
+import { ToastController } from '@ionic/angular'
 import { DiaryEntry } from '../model/diary-entry'
 import { DiaryService } from '../shared-services/diary/diary.service'
-import { LogfileService } from '../shared-services/logfile/logfile.service'
-
-import { FilesystemDirectory, Plugins } from '@capacitor/core'
-const { Filesystem, Share } = Plugins
-
-import JSZip from 'jszip'
 
 @Component({
   selector: 'app-diary',
@@ -22,10 +15,7 @@ export class DiaryPage implements OnInit {
   constructor(
     private router: Router,
     private diaryService: DiaryService,
-    private toastController: ToastController,
-    private translateService: TranslateService,
-    private logfileService: LogfileService,
-    private platform: Platform
+    private toastController: ToastController
   ) {}
 
   ngOnInit() {}
@@ -36,38 +26,9 @@ export class DiaryPage implements OnInit {
     )
   }
 
-  /**
-   * @description Generates a .zip File containing the latest diary and logfile data and exports it via the Capacitor `Share` plugin
-   */
   async exportDiary() {
     try {
-      const diaryData = await this.diaryService.exportDiary()
-      const logfileData = await this.logfileService.exportLog()
-
-      const timestamp = new Date()
-        .toISOString()
-        .replace(/:/g, '-')
-        .split('.')[0]
-
-      const zip = new JSZip()
-      zip.file(`diary_${timestamp}.csv`, diaryData)
-      zip.file(`log_${timestamp}.csv`, logfileData)
-
-      const data = await zip.generateAsync({ type: 'base64' })
-      const fileResult = await Filesystem.writeFile({
-        data,
-        path: `SIMPORT_export_${timestamp}.zip`,
-        directory: FilesystemDirectory.ExternalStorage,
-      })
-
-      if (this.platform.is('android')) {
-        await Share.requestPermissions()
-      }
-
-      Share.share({
-        title: this.translateService.instant('diary.exportFileName'),
-        url: fileResult.uri,
-      })
+      await this.diaryService.exportDiary()
     } catch (e) {
       const toast = await this.toastController.create({
         message: e.message,

--- a/src/app/shared-services/diary/diary.service.spec.ts
+++ b/src/app/shared-services/diary/diary.service.spec.ts
@@ -1,6 +1,7 @@
 import { TestBed } from '@angular/core/testing'
 import { APP_TEST_IMPORTS } from 'src/app/app.declarations'
 import { SqliteService } from '../db/sqlite.service'
+import { TrajectoryService } from '../trajectory/trajectory.service'
 
 import { DiaryService } from './diary.service'
 
@@ -10,7 +11,7 @@ describe('DiaryService', () => {
   beforeEach(() => {
     TestBed.configureTestingModule({
       imports: APP_TEST_IMPORTS,
-      providers: [SqliteService],
+      providers: [TrajectoryService, SqliteService],
     })
     service = TestBed.inject(DiaryService)
   })

--- a/src/app/shared-services/diary/diary.service.ts
+++ b/src/app/shared-services/diary/diary.service.ts
@@ -4,6 +4,12 @@ import { DiaryEntry } from 'src/app/model/diary-entry'
 import { SqliteService } from '../db/sqlite.service'
 import { Platform } from '@ionic/angular'
 import { TranslateService } from '@ngx-translate/core'
+import JSZip from 'jszip'
+import { FilesystemDirectory, Plugins } from '@capacitor/core'
+import { Device } from '@ionic-native/device'
+import { LogfileService } from '../logfile/logfile.service'
+
+const { Filesystem: filesystem, Share: share } = Plugins
 
 @Injectable({
   providedIn: 'root',
@@ -12,7 +18,8 @@ export class DiaryService {
   constructor(
     private dbService: SqliteService,
     private platform: Platform,
-    private translateService: TranslateService
+    private translateService: TranslateService,
+    private logfileService: LogfileService
   ) {}
 
   async getDiary(): Promise<DiaryEntry[]> {
@@ -49,10 +56,42 @@ export class DiaryService {
   }
 
   /**
-   *
-   * @returns diary as csv string
+   * @description Generates a .zip File containing the latest diary and logfile data and exports it via the Capacitor `Share` plugin
    */
   async exportDiary() {
+    const diaryData = await this.createDiaryExport()
+    const logfileData = await this.logfileService.exportLog()
+    const timestamp = new Date().toISOString().replace(/:/g, '-').split('.')[0]
+    const zip = new JSZip()
+
+    zip.file(`diary_${timestamp}.csv`, diaryData)
+    zip.file(`log_${timestamp}.csv`, logfileData)
+
+    const data = await zip.generateAsync({ type: 'base64' })
+    const directory = this.isAndroid10OrAbove()
+      ? FilesystemDirectory.Documents
+      : FilesystemDirectory.ExternalStorage
+
+    const fileResult = await filesystem.writeFile({
+      data,
+      path: `SIMPORT_export_${timestamp}.zip`,
+      directory,
+    })
+
+    if (this.platform.is('android')) {
+      await share.requestPermissions()
+    }
+
+    share.share({
+      title: this.translateService.instant('diary.exportFileName'),
+      url: fileResult.uri,
+    })
+  }
+
+  /**
+   * @returns diary as csv string
+   */
+  private async createDiaryExport(): Promise<string> {
     try {
       const diary = await this.getDiary()
 
@@ -72,5 +111,14 @@ export class DiaryService {
       )
       throw new Error(errorMessage)
     }
+  }
+
+  private isAndroid10OrAbove(): boolean {
+    if (this.platform.is('android')) {
+      const osVersion = parseInt(Device.version, 10) || 0
+      // 'always-allow' exists since OS-version 10 = API-level 29
+      return osVersion >= 10
+    }
+    return false
   }
 }

